website/server/controllers/api-v3/challenges.js

```diff
@@ -386,39 +387,31 @@ api.getUserChallenges = {
       query.categories = { $elemMatch: { slug: {$in: categorySlugs} } };
     }
 
-    let mongoQuery = Challenge.find(query)
-      .sort('-createdAt');
-
+    // Ensure that official challenges are always first
+    let mongoQuery = createChallengeQuery(query);
     if (page) {
       mongoQuery = mongoQuery
-        .limit(CHALLENGES_PER_PAGE)
-        .skip(CHALLENGES_PER_PAGE * page);
+        .skip(CHALLENGES_PER_PAGE * page)
+        .limit(CHALLENGES_PER_PAGE);
     }
 
     // see below why we're not using populate
     // .populate('group', basicGroupFields)
     // .populate('leader', nameFields)
-    const challenges = await mongoQuery.exec();
-
-
-    let resChals = challenges.map(challenge => challenge.toJSON());
-
-    resChals = _.orderBy(resChals, [challenge => {
-      return challenge.categories.map(category => category.slug).includes('habitica_official');
-    }], ['desc']);
+    let challenges = await mongoQuery.exec();
 
     // Instead of populate we make a find call manually because of https://github.com/Automattic/mongoose/issues/3833
-    await Promise.all(resChals.map((chal, index) => {
+    await Promise.all(challenges.map((chal, index) => {
       return Promise.all([
         User.findById(chal.leader).select(nameFields).exec(),
         Group.findById(chal.group).select(basicGroupFields).exec(),
       ]).then(populatedData => {
-        resChals[index].leader = populatedData[0] ? populatedData[0].toJSON({minimize: true}) : null;
-        resChals[index].group = populatedData[1] ? populatedData[1].toJSON({minimize: true}) : null;
+        challenges[index].leader = populatedData[0] ? populatedData[0].toJSON({minimize: true}) : null;
```

Comment:
is this working correctly? The reason why we first serialized the challenges in plain json objects and then applied the leader and group data is to avoid mongoose removing the changes since they aren't strings as in the schema

Reply:
I think so. aggregate() returns plain JSON objects, so we don’t have to call toJSON() on the results.

Reply:
Hmm does this means defaults aren't applied as well by Mongoose

Reply:
That’s a good question. I can check it out, if you want.

Reply:
Yes that would be useful, thanks. In fact if a plain object is returned we'd have to go back to the server side sorting because we rely on the data from the db passing through mongoose to be sure private data is not shared, defaults are applied correctly, etc...

Reply:
I'm looking at this right now, but I'm not exactly sure what I should be looking for. Is there something in particular I can do to check whether defaults are being applied?

I'm also not super sure what the issue is regarding the plain object. The Mongoose [docs](https://mongoosejs.com/docs/api.html#aggregate_Aggregate) say that a plain object is returned instead of a document "since any shape of document can be returned". I think the data still passes through Mongoose normally.